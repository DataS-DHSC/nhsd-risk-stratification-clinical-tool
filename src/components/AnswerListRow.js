import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import SectionAnswersSummary from './SectionAnswersSummary';

const filterSectionsWithAllDefaults = (sections, data) =>
  sections.filter((section) =>
    Object.keys(section.fields).some(
      (field) =>
        data[`answers.${field}`] !== section.fields[field].default ||
        section.fields[field].default === null
    )
  );

const AnswerListRow = ({ pageTemplate, data, formBaseUrl }) => {
  const { label, sections, formId } = pageTemplate;
  const url = `${formBaseUrl}/${formId}`;

  // Filter sections that have default values for ALL their fields
  const filteredSections = filterSectionsWithAllDefaults(sections, data);

  return (
    <>
      {filteredSections.length > 0 && (
        <h2
          id={`answer-list-${label.toLowerCase().replace(/ /g, '-')}-heading`}
        >
          {label}
        </h2>
      )}
      {filteredSections.map((pageSection) => (
        <Fragment key={`${label}-${pageSection.label}`}>
          {pageSection.label !== '' && (
            <h3
              id={`answer-list-${pageSection.label
                .toLowerCase()
                .replace(/ /g, '-')}-heading`}
            >
              {pageSection.label}
            </h3>
          )}
          <SectionAnswersSummary
            pageSection={pageSection}
            data={data}
            pageUrl={url}
          />
        </Fragment>
      ))}
    </>
  );
};

AnswerListRow.propTypes = {
  pageTemplate: PropTypes.shape().isRequired, // TODO: rename to 'form' to be consistent with FormGenerator
  data: PropTypes.shape().isRequired,
  formBaseUrl: PropTypes.string.isRequired,
};

export default AnswerListRow;
