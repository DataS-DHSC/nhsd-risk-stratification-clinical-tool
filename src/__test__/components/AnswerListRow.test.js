import React from 'react';
import { render } from '@testing-library/react';
import AnswerListRow from '../../components/AnswerListRow';

const testForm = (defaults) => ({
  label: 'form label',
  formId: 'form7b',
  sections: [
    {
      label: '',
      fields: {
        field1: {
          label: 'page1field1Label',
          type: 'text-input',
          default: defaults.field1,
        },
        field2: {
          label: 'page1field2Label',
          type: 'text-input',
          default: defaults.field2,
        },
      },
    },
  ],
});

const setup = ({ data, defaults }) =>
  render(
    <AnswerListRow
      pageTemplate={testForm(defaults)}
      data={data}
      formBaseUrl="https://example.com/prefix"
    />
  );

describe('AnswerListRow', () => {
  it('Should display section if default is null', () => {
    const component = setup({
      data: {
        'answers.field1': 'field1 should show in snapshot',
        'answers.field2': 'field2 should show in snapshot',
      },
      defaults: {
        field1: null,
        field2: null,
      },
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('Should display section if default is not null and input is not equal to default', () => {
    const component = setup({
      data: {
        'answers.field1': 'field1 should show in snapshot',
        'answers.field2': 'field2 should show in snapshot',
      },
      defaults: {
        field1: 'field1 default',
        field2: 'field2 default',
      },
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('Should not display section if default is not null and input is equal to default', () => {
    const component = setup({
      data: {
        'answers.field1': 'field1 default',
        'answers.field2': 'field2 default',
      },
      defaults: {
        field1: 'field1 default',
        field2: 'field2 default',
      },
    });

    expect(component.asFragment()).toMatchSnapshot();
  });
});
