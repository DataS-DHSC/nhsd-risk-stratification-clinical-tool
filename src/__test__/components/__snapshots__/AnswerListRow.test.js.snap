// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerListRow Should display section if default is not null and input is not equal to default 1`] = `
<DocumentFragment>
  <h2
    id="answer-list-form-label-heading"
  >
    form label
  </h2>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-field1"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        page1field1Label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        field1 should show in snapshot
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change page1field1Label from field1 should show in snapshot"
          class="nhsuk-button change-button"
          formaction="https://example.com/prefix/form7b"
          id="change-button-field1"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-field2"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        page1field2Label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        field2 should show in snapshot
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change page1field2Label from field2 should show in snapshot"
          class="nhsuk-button change-button"
          formaction="https://example.com/prefix/form7b"
          id="change-button-field2"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`AnswerListRow Should display section if default is null 1`] = `
<DocumentFragment>
  <h2
    id="answer-list-form-label-heading"
  >
    form label
  </h2>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-field1"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        page1field1Label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        field1 should show in snapshot
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change page1field1Label from field1 should show in snapshot"
          class="nhsuk-button change-button"
          formaction="https://example.com/prefix/form7b"
          id="change-button-field1"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-field2"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        page1field2Label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        field2 should show in snapshot
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change page1field2Label from field2 should show in snapshot"
          class="nhsuk-button change-button"
          formaction="https://example.com/prefix/form7b"
          id="change-button-field2"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`AnswerListRow Should not display section if default is not null and input is equal to default 1`] = `<DocumentFragment />`;
